rfcs/20210519-single-pass-bnorm/README.md

```diff
@@ -0,0 +1,126 @@
+# Proposal to calculate mean & variance in batchnorm in single pass
+
+## Requester
+GPU Enabling team
+
+## Motivation
+Using new formula for calculating variance will result in up to 33% performance
+improvement in batchnorm fwd training operation. 
+
+Currently bnorm performance is far below projections. Projections are based on
+assumption that bnorm reads input tensor once and writes result once. Current
+implementation requires bnorm to read input tensor 3 times. Proposed change
+will decrease it to 2 reads. There is no known generic algorithm to decrease it
+to 1 read, although such efficiency is possible for special cases where input
+tensor is small enough.
+
+## Proposal
+Use the following formula to calculate variance:
+```
+Var(X) = E(X^2) - (E(X))^2
+```
+where X = input tensor, E = average
+
+This formula allows hardware to calculate both mean and variance in single pass
+over input tensor.
+For reference, the canonical formula is:
+```
+Var(X) = E((X - Mean(X))^2)
+```
+The benefit of new formula is 25-33% performance improvement. Batchnorm kernel
+is memory-bound, with new formula it will need to read input tensor 2 times 
+(1: mean&var, 2: normalize) instead of 3 times (1: mean, 2: var, 3: normalize)
+
+This formula will be used whenever batchnorm kernel calculates mean and 
+variance, which means forward propagation in training. However, with 
+non-standard combination of flags, oneDNN's batchnorm implementaton may be
+configured to calculate those two values in inference as well.
+
+## Design
+Algorithm for calculating variance will be implemented as:
+```
+Sum = 0
+SumOfSquares = 0
+for each x in X:
+	Sum += x
+	SumOfSquares += (x*x)
+Variance = SumOfSquares/N - (Sum/N)^2
+if Variance < 0 then Variance = 0
+```
+where N = number of items in given channel in tensor.
+
+In the formula there's a subtraction of two large values which is vulnerable
+to catastrophic cancellation. Rounding errors in the two big numbers may be
+large in relation to result of the formula. In order to minimize rounding
+errors:
+- Sum and SumOfSquares will be stored in fp32 precision, regardless of input
+	data type
+- Kahan summation algorithm will be used
+
+The last line, "if var < 0 then var = 0" is there to make sure rounding errors
+won't cause variance to go negative. Further in batchnorm calculations square
+root is taken from variance. Negative value would result in NaN.
+
+New formula should be considered experimental. By default the old one will be
+used. There needs to be a way for user to choose formula. It is not decided yet
+how exactly user will toggle it.
+
+### Enable through environmental variable
+OneDNN will read env var `DNNL_BNORM_ONE_PASS`. If not set or equal to zero, 
+old formula will be used.
+
+Pros:
+- API/ABI compatibility
+- Simplest to implement and use
+- Clearly defined default state
+- If needed, may be kept confidential and not disclosed to external customes 
```

Comment:
onednn is an open source project, there is no reason to keep it secret, optimizations are vendor-agnostic.